src/source.rs

```diff
@@ -89,10 +88,40 @@ pub trait Locker: fmt::Debug {
   }
 }
 
-#[derive(Debug, Clone, PartialEq)]
-pub struct ResolveResult {
-  pub specifier: ModuleSpecifier,
-  pub kind: ModuleKind,
+#[derive(Debug)]
+pub enum ResolveResponse {
+  Amd(ModuleSpecifier),
+  CommonJs(ModuleSpecifier),
+  Err(Error),
+  Esm(ModuleSpecifier),
```

Comment:
Why is `Esm` separate from `Specifier`?

Reply:
`Specifier` indicates the resolver isn't really doing anything except resolving the specifier, where as `Esm` is asserting it is specifically an ESM module.  Current behaviour is that `Specifier` always gets assigned `ModuleKind::Esm` but we may want to change that in the future, or potentially perform further analysis on the module to determine the module kind.

Reply:
Ok, this still confuses me, but let's do it.

Reply:
@kitsonk could you add this info as doc comments?